sql: add support for optional frame exclusion

yuzefovich opened the pull request:
Adds support for optional frame exclusion clause in the specification
of window frames. This feature is described in
https://www.postgresql.org/docs/12/sql-expressions.html.

Whenever a non-default frame exclusion is present, it forces the
execution to fall back to a naive quadratic approach. This means
that for every row in a window frame we will compute the window
function from scratch, without making use of any previous computations.
Additionally, the presence of a non-default frame exclusion breaks
necessary assumptions for utilizing a sliding window approach for
min, max, sum, and avg functions, so those also fall back to a naive
quadratic approach.

Closes: #27100.

Release note (sql change): CockroachDB now supports optional frame
exclusion clause in the specification of window frames.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/28262)
<!-- Reviewable:end -->

yuzefovich commented:
Please ignore the first commit here - it is in a separate PR.

knz reviewed (commented):
Unfortunately today I only have time for a casual review. At that level the overall structure is ok but I really can't comment on correctness. I will be able to do a more thorough review in a few day (unless you find another reviewer).

There is a pretty-print error though, I'm pretty sure.

> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LJIlin92EE3NQ1f0bRR:bom7iew)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LJIlHBo2TkhQof0F_D-:by6ksau) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
> ```Go
> 		)
> 	}
> 	if wf.Exclusion != nil {
> ```

Are bounds mutually exclusive with an exclusion clause? (Casual inspection suggest they're not)

If not the code as you wrote it will print the wrong output (you're overriding `d`). 
If they are mutually exclusive please use `else if`

Also, make a test to illustrate this.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
This PR is of least priority (I haven't thought of a use case for this frame exclusion), so no rush with the review. It depends on #28244,  I definitely want to merge that one first.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LJKwY9g4rSIY1NXwPI1:b-bjslii)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LJKj71U3gaM_9RcEG6U:b-9u8n9b) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Are bounds mutually exclusive with an exclusion clause? (Casual inspection suggest they're not)

If not the code as you wrote it will print the wrong output (you're overriding `d`). 
If they are mutually exclusive please use `else if`

Also, make a test to illustrate this.
</blockquote></details>

You're right - they are not mutually exclusive. But I'm not sure what the correct way to "append" information about frame exclusion and how to test it - I welcome any hints.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
@knz @nvanbenschoten @jordanlewis It is now ready for review, please take a look when you get a chance.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LK1wcr_4hVxxvrav8kv:bqf7rcm)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LK1wHDf3LxcnxVosQxg:b-2hsyc3) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

You're right - they are not mutually exclusive. But I'm not sure what the correct way to "append" information about frame exclusion and how to test it - I welcome any hints.
</blockquote></details>

@knz what is the correct way to include information about frame exclusion? And how would I test it?


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 15 of 15 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LK2-SCq1ITA3tVSt3fy:b-avexmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

@knz what is the correct way to include information about frame exclusion? And how would I test it?
</blockquote></details>

simply: `d = pretty.ConcatSpace(d, p.Doc(wf.Exclusion))`

together with my other suggested changes
___
*[pkg/sql/sem/tree/pretty.go, line 713 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-Q5A1PwMcdJEdtzn:-LK2-Q5A1PwMcdJEdtzo:btm5pn) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/pretty.go#L713)):*
> ```Go
> }
> 
> func (node *WindowFrameExclusion) doc(p *PrettyCfg) pretty.Doc {
> ```

Remove this function.
___
*[pkg/sql/sem/tree/select.go, line 771 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-6hfEEYhYZaxiHNW:-LK2-6hg6D6bP_DWC7g0:b-ib9vkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/select.go#L771)):*
> ```Go
> 	switch *node {
> 	case ExcludeCurrentRow:
> 		ctx.WriteString(" EXCLUDE CURRENT ROW")
> ```

Factor the `"EXCLUDE"` prefix to the top of the `switch`. Do not prepend a space here.
___
*[pkg/sql/sem/tree/select.go, line 804 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-La3CMf-TK31HPa-:-LK2-La3CMf-TK31HPa0:bw4x22z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/select.go#L804)):*
> ```Go
> 	}
> 	if node.Exclusion != nil {
> 		ctx.FormatNode(node.Exclusion)
> ```

add `ctx.WriteByte(' ')` before the call to FormatNode here.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LK21dnF-KsIf69Fa8Xe:bufhbbg) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

simply: `d = pretty.ConcatSpace(d, p.Doc(wf.Exclusion))`

together with my other suggested changes
</blockquote></details>

Thanks, this makes sense.

How would I test this pretty printing? Do I need to create a SQL file with a query in `pkg/sql/sem/tree/testdata/pretty/`? How would I run it?
___
*[pkg/sql/sem/tree/pretty.go, line 713 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-Q5A1PwMcdJEdtzn:-LK20TQfCHxTv4X9vyes:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/pretty.go#L713)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Remove this function.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 771 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-6hfEEYhYZaxiHNW:-LK20whN9xsjDQnk2f3k:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/select.go#L771)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Factor the `"EXCLUDE"` prefix to the top of the `switch`. Do not prepend a space here.
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 804 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LK2-La3CMf-TK31HPa-:-LK20sIxBT6wMj3uHyyi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3da0a45c886794c756d30a852587cbfe61e68ec0/pkg/sql/sem/tree/select.go#L804)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

add `ctx.WriteByte(' ')` before the call to FormatNode here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
I think I currently have a bug with some non-aggregate window functions (like `row_number` and `rank`) - at the moment, they simply ignore frame exclusion. I will fix this shortly and will expand the test coverage.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LK24pgx-X51GpidM64e:b-to13ka)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LK2A5NKBtAVAR3F0JK6:b-fkmfp5) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Thanks, this makes sense.

How would I test this pretty printing? Do I need to create a SQL file with a query in `pkg/sql/sem/tree/testdata/pretty/`? How would I run it?
</blockquote></details>

Make an example in that directory yes, then run `make test PKG=./pkg/sql/sem/tree TESTFLAGS=-rewrite-pretty`

then add the new files (inspect them manually to see if they are to your liking)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Actually I take that back. Only `first_value`, `last_value` and `nth_value` probably should take frame exclusion into consideration when computing window functions (confirmed this with PG 11 beta 2), and those are handled correctly. I added tests for those three (tests are not very diverse though because it's hard to make deterministic ones while considering boundaries of peer groups).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LK2LZEb5-YVspTvX1TU:bwyymyo)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LK2I2L11l7lqtSEFBTg:bcr5s31) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make an example in that directory yes, then run `make test PKG=./pkg/sql/sem/tree TESTFLAGS=-rewrite-pretty`

then add the new files (inspect them manually to see if they are to your liking)
</blockquote></details>

Got it, thanks.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
@knz @jordanlewis @nvanbenschoten please take a look at this guy :)

knz reviewed (commented):
> Reviewed 12 of 23 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LcHXqBz4y11OVA2IXVw:br9johy) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Got it, thanks.
</blockquote></details>

I think `pretty.Stack` will look better than `ConcatSpace`. Can you try it?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 23 of 23 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @yuzefovich)

___
*[pkg/sql/distsqlpb/processors.go, line 222 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSwXPNDzKJ-Gc8e5z-:-LcSwXPO1FB8wQ6G1ItC:bd3apxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/distsqlpb/processors.go#L222)):*
> ```Go
> 			// that is non-nil.
> 			var exclusion WindowerSpec_Frame_Exclusion
> 			spec.Exclusion = &exclusion
> ```

nit: `spec.Exclusion = new(WindowerSpec_Frame_Exclusion)` is slightly more explicit if all we're looking for is an allocation.
___
*[pkg/sql/distsqlpb/processors.proto, line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSybNM8HwSf7tDqmRs:-LcSybNM8HwSf7tDqmRt:b-h5wyt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/distsqlpb/processors.proto#L998)):*
> ```protobuf
>     optional Mode mode = 1 [(gogoproto.nullable) = false];
>     optional Bounds bounds = 2 [(gogoproto.nullable) = false];
>     optional Exclusion exclusion = 3;
> ```

Same point as I make below. I'd strongly consider making this non-nullable and adding a variant to in the enum to indicate an unspecified frame exclusion.
___
*[pkg/sql/sem/builtins/window_builtins.go, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSz4Vp1YBEmGwwOuIa:-LcSz4Vp1YBEmGwwOuIb:b-e3k4yn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/builtins/window_builtins.go#L221)):*
> ```Go
> 		// Note that we ignore frame exclusion here because it must not be
> 		// present - framableAggregateWindowFunc itself handles the case of non
> 		// default frame exclusion.
> ```

Consider asserting that.
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSzhEK0_-nA1tc4NiI:-LcSzhEK0_-nA1tc4NiJ:b-5s9151) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/builtins/window_frame_builtins.go#L273)):*
> ```Go
> 		return nil, err
> 	}
> 	if !wfr.DefaultFrameExclusion() {
> ```

I don't have a great memory of how this all works, but why use a `slidingWindowSumFunc` if we're going to fall back to the naive quadratic approach anyway? Is this a case where we should be detecting the non-default frame exclusion at the point where we construct the window function instead of detecting it here?
___
*[pkg/sql/sem/tree/select.go, line 798 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSx0h9Bh3l8if1G0pq:-LcSx0hA-2f3YYLt3nDI:bdvjirg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/tree/select.go#L798)):*
> ```Go
> 	Mode      WindowFrameMode       // the mode of framing being used
> 	Bounds    WindowFrameBounds     // the bounds of the frame
> 	Exclusion *WindowFrameExclusion // optional frame exclusion
> ```

Instead of making this nullable to indicate an unspecified frame exclusion, I'd suggest you consider making the default value of `WindowFrameExclusion` indicate no frame exclusion specified. There's no reason to force an allocation here and you're bound to run into issues with reference equality vs. value equality at some point.


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
CC @justinj

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @nvanbenschoten)

___
*[pkg/sql/distsqlpb/processors.go, line 222 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSwXPNDzKJ-Gc8e5z-:-LcXrDK87Lzixdz5OvLw:blbhgq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/distsqlpb/processors.go#L222)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: `spec.Exclusion = new(WindowerSpec_Frame_Exclusion)` is slightly more explicit if all we're looking for is an allocation.
</blockquote></details>

Yes, that's what I was looking for and forgot about `new`.
___
*[pkg/sql/distsqlpb/processors.proto, line 998 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSybNM8HwSf7tDqmRs:-LcXrLjx19AldcOQoWn5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/distsqlpb/processors.proto#L998)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same point as I make below. I'd strongly consider making this non-nullable and adding a variant to in the enum to indicate an unspecified frame exclusion.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/window_builtins.go, line 221 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSz4Vp1YBEmGwwOuIa:-LcXrtv94uA0pT8ZXuZ6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/builtins/window_builtins.go#L221)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider asserting that.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 273 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSzhEK0_-nA1tc4NiI:-LcXs_I759J15q90hSAL:bx6wk8t) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/builtins/window_frame_builtins.go#L273)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I don't have a great memory of how this all works, but why use a `slidingWindowSumFunc` if we're going to fall back to the naive quadratic approach anyway? Is this a case where we should be detecting the non-default frame exclusion at the point where we construct the window function instead of detecting it here?
</blockquote></details>

So what's happening is that here (https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/sem/builtins/aggregate_builtins.go#L346) we replace certain aggregates (like `sum`) when used as window functions with optimized versions specifically. I believe it happens once, on the cluster startup. At that point we don't know anything about frame exclusion.

We could equip these sliding window optimized functions with a corresponding to it naive quadratic `framableAggregateWindowFunction`s that will take of the computation. That will reduce the two instances of copy/paste but will probably make the code more complicated. I can try if you think it'll be better.
___
*[pkg/sql/sem/tree/pretty.go, line 687 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LJIlHBm6TQ6oxevyQ7s:-LcXwExsCv-jkcWILH9y:b-14ho7x) ([raw file](https://github.com/cockroachdb/cockroach/blob/018d9fff6a9cc8cd8ba226b602e08ac6d9f9a675/pkg/sql/sem/tree/pretty.go#L687)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think `pretty.Stack` will look better than `ConcatSpace`. Can you try it?
</blockquote></details>

I don't have a strong preference. Let me know which one you prefer.
___
*[pkg/sql/sem/tree/select.go, line 798 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LcSx0h9Bh3l8if1G0pq:-LcXr6P16f63k_Q9PUE9:b-muxp26) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd1382ac2744e8f2333fae51860a3f2e067650ba/pkg/sql/sem/tree/select.go#L798)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Instead of making this nullable to indicate an unspecified frame exclusion, I'd suggest you consider making the default value of `WindowFrameExclusion` indicate no frame exclusion specified. There's no reason to force an allocation here and you're bound to run into issues with reference equality vs. value equality at some point.
</blockquote></details>

Thanks, I like this suggestion, done.


<!-- Sent from Reviewable.io -->

justinj commented:
soft request to avoid adding new features to window functions until we have the optimizer up to par with the HP :)

yuzefovich committed:
sql: add support for optional frame exclusion

Adds support for optional frame exclusion clause in the specification
of window frames. This feature is described in
https://www.postgresql.org/docs/12/sql-expressions.html.

Whenever a non-default frame exclusion is present, it forces the
execution to fall back to a naive quadratic approach. This means
that for every row in a window frame we will compute the window
function from scratch, without making use of any previous computations.
Additionally, the presence of a non-default frame exclusion breaks
necessary assumptions for utilizing a sliding window approach for
min, max, sum, and avg functions, so those also fall back to a naive
quadratic approach.

Release note (sql change): CockroachDB now supports optional frame
exclusion clause in the specification of window frames.

yuzefovich commented:
Now that work on window functions in the optimizer is done (at least, for the moment), let's merge this. PTAL.

justinj reviewed (commented):
Would you mind adding some tests to `optbuilder/testdata/window` and `execbuilder/testdata/window`? Just to make sure the information is getting threaded through OK. Besides that opt changes <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 28 of 29 files at r5, 2 of 2 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LlX9Tyr0xwTe94pzWp5:bhkiwlv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/window, line 3006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LlXFqdT8EYxlrHZdjYT:-LlXFqdU2UlTg6xlDzs1:byldndr) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbaaa51393220fa56cbf644898cadd2cce164896/pkg/sql/logictest/testdata/logic_test/window#L3006)):*
> ```
> 
> query RTR
> SELECT price, array_agg(price) OVER w, sum(price) OVER w FROM products WINDOW w AS (ORDER BY price RANGE UNBOUNDED PRECEDING EXCLUDE CURRENT ROW) ORDER BY price
> ```

nit: consider sqlfmting these queries


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Added, thanks for the suggestion.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LlblakI-vg1WbrOCfoS:bl3aq0s)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @knz, and @nvanbenschoten)

___
*[pkg/sql/logictest/testdata/logic_test/window, line 3006 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LlXFqdT8EYxlrHZdjYT:-Llbl__ADqXMG8bvHrOR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bbaaa51393220fa56cbf644898cadd2cce164896/pkg/sql/logictest/testdata/logic_test/window#L3006)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

nit: consider sqlfmting these queries
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Friendly ping.

knz reviewed (commented):
I don't have enough knowledge of frame exclusion modes to properly assess the correctness of the feature, but the code is overall consistent with out best practices and the testing is adequate. I left a few minor comments but this is code-wise good to go.

I would suggest two things:

1) extend the commit message with a link to the pg docs where this feature is explained (if pg supports it now) otherwise docs from some other sql databases that do, so the doc team and our future selves have some reference to look at.

2) provide some commentary in the commit message about the algorithmic costs, so that we can build a mental model of what to expect in memory/cpu usage.

> Reviewed 12 of 29 files at r5, 1 of 2 files at r6, 10 of 10 files at r7.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LmOpQM5CPvTFbjhqEpp:b-24rr8r)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @nvanbenschoten, and @yuzefovich)

___
*[pkg/sql/distsqlpb/processors.go, line 89 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOhI3a4q5UULfl2h6J:-LmOhI3a4q5UULfl2h6K:bmf03y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L89)):*
> ```Go
> 		*spec = WindowerSpec_Frame_EXCLUDE_TIES
> 	default:
> 		panic("unexpected WindowerFrameExclusion")
> ```

`panic(errors.NewAssertionErrorf(...))`

but really this function should return `error`
___
*[pkg/sql/distsqlpb/processors.go, line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOhPPM9fIYNenYOyLK:-LmOhPPN5LA9w8byIS9i:bizxaqx) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L207)):*
> ```Go
> // offset expressions if present in the frame.
> func (spec *WindowerSpec_Frame) InitFromAST(f *tree.WindowFrame, evalCtx *tree.EvalContext) error {
> 	spec.Mode.initFromAST(f.Mode)
> ```

and you should test the error return here
___
*[pkg/sql/distsqlpb/processors.go, line 253 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOnZdjDsk9scxCHVSh:-LmOnZdkCjJWwQFAOQI1:b-uuwo2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L253)):*
> ```Go
> 		return tree.ExcludeTies
> 	default:
> 		panic("unexpected WindowerSpec_Frame_Exclusion")
> ```

`AssertionFailedf`
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 111 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOoR5I0kfBGVT2wj5Q:-LmOoR5I0kfBGVT2wj5R:b-8rqtb2) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/builtins/window_frame_builtins.go#L111)):*
> ```Go
> 
> 	if !wfr.DefaultFrameExclusion() {
> 		// We cannot use sliding window approach because we have frame exclusion
> ```

nit: "we cannot use _a_ sliding window approach because we have _a_ frame exclusion clause"
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOoWL34CJA9orY50-C:-LmOoWL34CJA9orY50-D:b5kau6c) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/builtins/window_frame_builtins.go#L113)):*
> ```Go
> 		// We cannot use sliding window approach because we have frame exclusion
> 		// clause - some rows will be in and out of the frame which breaks the
> 		// necessary assumption, so we fallback to naive quadratic approach.
> ```

"to a naive"
___
*[pkg/sql/sem/tree/select.go, line 884 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOom_00-l7P4YrT-MO:-LmOom_00-l7P4YrT-MP:b-1axc7j) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/select.go#L884)):*
> ```Go
> 		ctx.WriteString("TIES")
> 	default:
> 		panic(fmt.Sprintf("unhandled case: %d", node))
> ```

`errors.AssertionFailedf`
___
*[pkg/sql/sem/tree/window_funcs.go, line 526 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOp0BgAzKrDZQ-50GS:-LmOp0BgAzKrDZQ-50GT:bkdpeex) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L526)):*
> ```Go
> // rows of the partition are inside of the window for each of the rows.
> func (wfr *WindowFrameRun) FullPartitionIsInWindow() bool {
> 	if wfr.Frame == nil || !wfr.NoFilter() || !wfr.DefaultFrameExclusion() {
> ```

Is it maybe worth putting this 3-component predicate into a method of `WindowFrameRun` and use it everywhere. The risk to make a mistake throughout the code is currently too high IMHO.
___
*[pkg/sql/sem/tree/window_funcs.go, line 583 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOow-HDaSJDPyL6ErZ:-LmOow-HDaSJDPyL6Er_:bj5pm0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L583)):*
> ```Go
> // the window frame of the current row.
> func (wfr *WindowFrameRun) isRowExcluded(idx int) bool {
> 	if wfr.Frame == nil || wfr.Frame.Exclusion == NoExclusion {
> ```

`if wfr.DefaultFrameExclusion() {`
___
*[pkg/sql/sem/tree/window_funcs.go, line 599 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOpAQb-A9IwgAQ-L17:-LmOpAQb-A9IwgAQ-L18:b-uuwo2q) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L599)):*
> ```Go
> 		return curRowFirstPeerIdx <= idx && idx < curRowFirstPeerIdx+curRowPeerGroupRowCount && idx != wfr.RowIdx
> 	default:
> 		panic("unexpected WindowFrameExclusion")
> ```

`AssertionFailedf`


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for the review!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LmPv7054azh8uQZpBoC:b-9qdoz3)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @justinj, @knz, and @nvanbenschoten)

___
*[pkg/sql/distsqlpb/processors.go, line 89 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOhI3a4q5UULfl2h6J:-LmPbktH2_MUWvhKbepM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L89)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`panic(errors.NewAssertionErrorf(...))`

but really this function should return `error`
</blockquote></details>

Done.
___
*[pkg/sql/distsqlpb/processors.go, line 207 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOhPPM9fIYNenYOyLK:-LmPblQJ543ZoBhhkICO:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L207)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

and you should test the error return here
</blockquote></details>

Done.
___
*[pkg/sql/distsqlpb/processors.go, line 253 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOnZdjDsk9scxCHVSh:-LmPbmHL-Kb8CmPCaF6w:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/distsqlpb/processors.go#L253)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`AssertionFailedf`
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 111 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOoR5I0kfBGVT2wj5Q:-LmPbxiU-nDkfOc5xllP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/builtins/window_frame_builtins.go#L111)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

nit: "we cannot use _a_ sliding window approach because we have _a_ frame exclusion clause"
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/window_frame_builtins.go, line 113 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOoWL34CJA9orY50-C:-LmPbyW8AhXTOHtM5GLH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/builtins/window_frame_builtins.go#L113)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

"to a naive"
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/select.go, line 884 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOom_00-l7P4YrT-MO:-LmPcNFR3pbsdgLinPKJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/select.go#L884)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`errors.AssertionFailedf`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/window_funcs.go, line 526 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOp0BgAzKrDZQ-50GS:-LmPs11z9CRLP3TpK1Fa:bag3odf) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L526)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Is it maybe worth putting this 3-component predicate into a method of `WindowFrameRun` and use it everywhere. The risk to make a mistake throughout the code is currently too high IMHO.
</blockquote></details>

I think this particular predicate is used only here, but your suggestion prompted me to think more about whether checking if the filter is present should be a part of it (as well as in other places), and I'm not sure anymore that the current version is correct. It's a very tricky stuff.
___
*[pkg/sql/sem/tree/window_funcs.go, line 583 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOow-HDaSJDPyL6ErZ:-LmPchY88C3y_JW8FSF0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L583)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`if wfr.DefaultFrameExclusion() {`
</blockquote></details>

Done.
___
*[pkg/sql/sem/tree/window_funcs.go, line 599 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOpAQb-A9IwgAQ-L17:-LmPtRqV5KD_vxCaCXOu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L599)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`AssertionFailedf`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Adding a do-not-merge label as a reminder for myself to think more about the filter clause.

yuzefovich reviewed (commented):
The check for whether a filter is present is not needed in many cases where I added it in this PR, so I removed those checks.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LmWeqb-DPEu0nGGYlQ-:bo61bu4)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @justinj, @knz, and @nvanbenschoten)

___
*[pkg/sql/sem/tree/window_funcs.go, line 526 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-LmOp0BgAzKrDZQ-50GS:-LmWdFqf-igY7DNFhozd:b2ztj8c) ([raw file](https://github.com/cockroachdb/cockroach/blob/99d72d8643b343f2c48823554014e80ca7e16868/pkg/sql/sem/tree/window_funcs.go#L526)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think this particular predicate is used only here, but your suggestion prompted me to think more about whether checking if the filter is present should be a part of it (as well as in other places), and I'm not sure anymore that the current version is correct. It's a very tricky stuff.
</blockquote></details>

Okay, so checking the filter here is not necessary.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
Unless someone objects, I'll merge this on Monday.

knz reviewed (approved):
Code LGTM. I'll trust you and other reviewers on the feature's correctness.

> Reviewed 1 of 29 files at r5, 14 of 14 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/28262#-:-LmZdZOj-17VQYd69CTs:bm7vz89)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks everyone for the review!

bors r+
